// poro/dist/cjs/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Riot = exports.RiotClient = exports.Cargo = exports.schemaMap = exports.CargoClient = exports.general = exports.DDRAGON_DOMAIN = exports.DataDragon = exports.RAW_CDRAGON_DOMAIN = exports.CDRAGON_DOMAIN = exports.CommunityDragon = void 0;
const tslib_1 = require("tslib");
var cdragon_1 = require("./riot/static/cdragon");
Object.defineProperty(exports, "CommunityDragon", { enumerable: true, get: function () { return cdragon_1.CommunityDragon; } });
Object.defineProperty(exports, "CDRAGON_DOMAIN", { enumerable: true, get: function () { return cdragon_1.CDRAGON_DOMAIN; } });
Object.defineProperty(exports, "RAW_CDRAGON_DOMAIN", { enumerable: true, get: function () { return cdragon_1.RAW_CDRAGON_DOMAIN; } });
var ddragon_1 = require("./riot/static/ddragon");
Object.defineProperty(exports, "DataDragon", { enumerable: true, get: function () { return ddragon_1.DataDragon; } });
Object.defineProperty(exports, "DDRAGON_DOMAIN", { enumerable: true, get: function () { return ddragon_1.DDRAGON_DOMAIN; } });
var general_1 = require("./riot/static/general");
Object.defineProperty(exports, "general", { enumerable: true, get: function () { return general_1.general; } });
var leaguepedia_1 = require("./leaguepedia");
Object.defineProperty(exports, "CargoClient", { enumerable: true, get: function () { return leaguepedia_1.CargoClient; } });
var schema_1 = require("./leaguepedia/schema");
Object.defineProperty(exports, "schemaMap", { enumerable: true, get: function () { return schema_1.schemaMap; } });
exports.Cargo = tslib_1.__importStar(require("./leaguepedia/types"));
var apis_1 = require("./riot/apis");
Object.defineProperty(exports, "RiotClient", { enumerable: true, get: function () { return apis_1.RiotClient; } });
tslib_1.__exportStar(require("./riot/apis/utils"), exports);
exports.Riot = tslib_1.__importStar(require("./types"));
