// poro/dist/cjs/src/riot/apis/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Division = exports.Tier = exports.Queue = exports.Region = exports.Platform = void 0;
var Platform;
(function (Platform) {
    Platform["BR"] = "BR1";
    Platform["EUNE"] = "EUN1";
    Platform["EUW"] = "EUW1";
    Platform["JP"] = "JP1";
    Platform["KR"] = "KR";
    Platform["LAN"] = "LA1";
    Platform["LAS"] = "LA2";
    Platform["NA"] = "NA1";
    Platform["OCE"] = "OC1";
    Platform["TR"] = "TR1";
    Platform["RU"] = "RU";
})(Platform || (exports.Platform = Platform = {}));
var Region;
(function (Region) {
    /** BR, LAN, LAS, NA, OCE */
    Region["AMERICAS"] = "AMERICAS";
    /** JP, KR */
    Region["ASIA"] = "ASIA";
    /** EUNE, EUW, RU, TR */
    Region["EUROPE"] = "EUROPE";
})(Region || (exports.Region = Region = {}));
var Queue;
(function (Queue) {
    Queue["RANKED_SOLO_5x5"] = "RANKED_SOLO_5x5";
    Queue["RANKED_TFT_PAIRS"] = "RANKED_TFT_PAIRS";
    Queue["RANKED_TFT"] = "RANKED_TFT";
    Queue["RANKED_FLEX_SR"] = "RANKED_FLEX_SR";
    Queue["RANKED_FLEX_TT"] = "RANKED_FLEX_TT";
})(Queue || (exports.Queue = Queue = {}));
var Tier;
(function (Tier) {
    Tier["CHALLENGER"] = "CHALLENGER";
    Tier["GRANDMASTER"] = "GRANDMASTER";
    Tier["MASTER"] = "MASTER";
    Tier["DIAMOND"] = "DIAMOND";
    Tier["EMERALD"] = "EMERALD";
    Tier["PLATINUM"] = "PLATINUM";
    Tier["GOLD"] = "GOLD";
    Tier["SILVER"] = "SILVER";
    Tier["BRONZE"] = "BRONZE";
    Tier["IRON"] = "IRON";
})(Tier || (exports.Tier = Tier = {}));
var Division;
(function (Division) {
    Division["I"] = "I";
    Division["II"] = "II";
    Division["III"] = "III";
    Division["IV"] = "IV";
})(Division || (exports.Division = Division = {}));
