// poro/dist/cjs/src/leaguepedia/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CargoClient = void 0;
const tslib_1 = require("tslib");
const axios_1 = tslib_1.__importDefault(require("axios"));
const axios_retry_1 = tslib_1.__importDefault(require("axios-retry"));
const generateUrl_1 = require("./generateUrl");
const schema_1 = require("./schema");
const LEAGUEPEDIA_BASE_URL = 'https://lol.fandom.com';
class CargoClient {
    constructor({ metadataPrefix } = {}) {
        this.axiosInstance = axios_1.default.create({
            baseURL: LEAGUEPEDIA_BASE_URL,
        });
        this.metadataPrefix = metadataPrefix !== null && metadataPrefix !== void 0 ? metadataPrefix : '';
        (0, axios_retry_1.default)(this.axiosInstance, {
            retryCondition(err) {
                var _a;
                const errCodes = ['ECONNRESET', 'ETIMEDOUT'];
                const errCode = (_a = err.code) !== null && _a !== void 0 ? _a : '';
                if (errCodes.includes(errCode)) {
                    return true;
                }
                return false;
            },
        });
    }
    spaceToUnderscore(obj) {
        // No index signature with a parameter of type 'string' was found on type '{}'.
        // https://stackoverflow.com/a/66406882/13151903
        const newObj = {};
        Object.entries(obj).forEach(([key, value]) => {
            newObj[key.replaceAll(' ', '_')] = value;
        });
        return newObj;
    }
    convert(obj, tables) {
        const newObj = {};
        tables.forEach((t) => {
            const schema = schema_1.schemaMap[t];
            Object.entries(obj).forEach(([key, value]) => {
                if (!(key in schema))
                    return;
                const defaultValue = schema[key];
                switch (typeof defaultValue) {
                    case 'boolean': {
                        // leaguepedia use bit(1) to store boolean
                        newObj[key] = typeof value === 'number' ? Boolean(value) : null;
                        break;
                    }
                    case 'number': {
                        const n = defaultValue ? parseFloat(value) : parseInt(value);
                        newObj[key] = isNaN(n) ? null : n;
                        break;
                    }
                    case 'string': {
                        newObj[key] = value ? String(value) : null;
                        break;
                    }
                    case 'object': {
                        if (defaultValue instanceof Date) {
                            newObj[key] = value ? new Date(value + 'Z') : null;
                        }
                        else {
                            newObj[key] = value.map((item) => String(item));
                        }
                        break;
                    }
                    default: {
                        newObj[key] = value;
                    }
                }
            });
        });
        return newObj;
    }
    addPrefixToMetadata(obj) {
        // No index signature with a parameter of type 'string' was found on type '{}'.
        // https://stackoverflow.com/a/66406882/13151903
        let newObj = {};
        if (this.metadataPrefix === '') {
            newObj = obj;
        }
        else {
            Object.entries(obj).forEach(([key, value]) => {
                if (key.startsWith('_')) {
                    newObj[this.metadataPrefix + key] = value;
                }
                else {
                    newObj[key] = value;
                }
            });
        }
        return newObj;
    }
    query(parameter) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const url = yield (0, generateUrl_1.generateUrl)(parameter);
            const res = yield this.axiosInstance.get(url);
            return Object.assign(Object.assign({}, res), { data: res.data
                    // `spaceToUnderscore` should be executed first to guarantee object shape is as same as schema defined in `schemaMap`
                    .map(this.spaceToUnderscore.bind(this))
                    .map((item) => this.convert(item, parameter.tables))
                    .map(this.addPrefixToMetadata.bind(this)) });
        });
    }
}
exports.CargoClient = CargoClient;
